avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java

```diff
@@ -71,4 +78,35 @@ protected Object deserialize(byte[] payload) throws SerializationException {
       throw new SerializationException("Error deserializing Avro message for id " + id, e);
     }
   }
+
+  private DatumReader getDatumReader(Schema writerSchema) {
+    if (useSpecificAvroReader) {
+      return new SpecificDatumReader(writerSchema, getReaderSchema(writerSchema));
+    } else {
+      return new GenericDatumReader(writerSchema);
+    }
+  }
+
+  private Schema getReaderSchema(Schema writerSchema) {
+    Schema readerSchema = targetSchemaCache.get(writerSchema);
+    if (readerSchema == null) {
+      Class targetClass = SpecificData.get().getClass(writerSchema);
```

Comment:
From my point of view the use case for this is to convert data from other compatible schemas for a given subject to the schema that is compiled into the running code. We test and work against this compiled in schema so then as long as all schemas written into kafka are compatible then we know it will still work for new and old data.

The code for [`getClass()`](https://github.com/apache/avro/blob/trunk/lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java#L150) looks at the namespace/name as you said from the schema and tries to lookup that class. So there will ever only be one reader schema per namespace/name, which could be for many different writer schemas.

For us subject is a 1:1 mapping with namespace/name so then the compatibility falls back to ensuring that all schemas in a subjected are **fully** compatible. We check this pre-merge for new schemas so we know they are consistent.

What we could do here to double check again is to check the readerSchema is compatible the first time it is found for a given namespace/name and raise an exception if it's not? Though that compatibility would be for every pair of schemas, so that would make it a bit more complicated. We would still only have to do it once per pair so probably better than eventually silently failing with a bad error and irrelevant message (which is what avro gives you I think).